kernel/crates/kernel_elfloader/src/lib.rs

```diff
@@ -193,3 +193,576 @@ where
+        data[56..58].copy_from_slice(&1u16.to_le_bytes());
+
+        // Program header: PT_LOAD with read-only
+        let ph_offset = 64;
+        data[ph_offset..ph_offset + 4].copy_from_slice(&1u32.to_le_bytes()); // PT_LOAD
+        data[ph_offset + 4..ph_offset + 8].copy_from_slice(&4u32.to_le_bytes()); // R only
+        data[ph_offset + 16..ph_offset + 24].copy_from_slice(&0x3000usize.to_le_bytes()); // vaddr
+        data[ph_offset + 24..ph_offset + 32].copy_from_slice(&0x3000usize.to_le_bytes()); // paddr
+        data[ph_offset + 32..ph_offset + 40].copy_from_slice(&0usize.to_le_bytes()); // filesz
+        data[ph_offset + 40..ph_offset + 48].copy_from_slice(&0x100usize.to_le_bytes()); // memsz
+        data[ph_offset + 48..ph_offset + 56].copy_from_slice(&0x1000usize.to_le_bytes()); // align
+
+        let data_slice = aligned_vec_as_bytes(&aligned_data, size);
+        let elf_file = ElfFile::try_parse(data_slice).unwrap();
+        let result = loader.load(elf_file);
+        assert!(result.is_ok());
+        let image = result.unwrap();
+        assert_eq!(image.executable_allocations().len(), 0);
+        assert_eq!(image.readonly_allocations().len(), 1);
+        assert_eq!(image.writable_allocations().len(), 0);
+    }
+
+    #[test]
+    fn test_load_elf_allocation_failure() {
+        let memory_api = MockMemoryApi::with_failing_allocation();
+        let mut loader = ElfLoader::new(memory_api);
+
+        let size = 64 + 56;
+        let mut aligned_data = create_aligned_vec(size);
+        let data = aligned_vec_as_bytes_mut(&mut aligned_data, size);
+        let header_data = create_minimal_elf_header();
+        data[..64].copy_from_slice(header_data.as_slice());
+
+        data[32..40].copy_from_slice(&64usize.to_le_bytes());
+        data[56..58].copy_from_slice(&1u16.to_le_bytes());
+
+        let ph_offset = 64;
+        data[ph_offset..ph_offset + 4].copy_from_slice(&1u32.to_le_bytes()); // PT_LOAD
+        data[ph_offset + 4..ph_offset + 8].copy_from_slice(&5u32.to_le_bytes()); // R+X
+        data[ph_offset + 16..ph_offset + 24].copy_from_slice(&0x1000usize.to_le_bytes()); // vaddr
+        data[ph_offset + 40..ph_offset + 48].copy_from_slice(&0x100usize.to_le_bytes()); // memsz
+        data[ph_offset + 48..ph_offset + 56].copy_from_slice(&0x1000usize.to_le_bytes()); // align
+
+        let data_slice = aligned_vec_as_bytes(&aligned_data, size);
+        let elf_file = ElfFile::try_parse(data_slice).unwrap();
+        let result = loader.load(elf_file);
+        assert!(matches!(result, Err(LoadElfError::AllocationFailed)));
+    }
+
+    #[test]
+    fn test_load_elf_with_tls_segment() {
+        let memory_api = MockMemoryApi::new();
+        let mut loader = ElfLoader::new(memory_api);
+
+        let size = 64 + 56;
+        let mut aligned_data = create_aligned_vec(size);
+        let data = aligned_vec_as_bytes_mut(&mut aligned_data, size);
+        let header_data = create_minimal_elf_header();
```

Comment:
This always seems so bloated and not nice. Maybe you can find a way to use `slice::align_to` in a way to easily create an aligned slice instead of having to go through `Vec`s and alignment? (Applies to this whole PR)

I imagine instead of the initial `[0_u8; 64]`, you could create a single (maybe even const) function like `<const N:usize> aligned_slice::<N>() -> [u8; N]` (not sure whether something like that works with ownership/lifetimes), that would start out with a `[0_u8; 2*N-1]` and then call `align_to(N)`, discarding the prefix and suffix and leaving you with a nicely aligned slice that you can use.

You have to figure out the details, but it seems nicer than the current approach.